src/tactic/lint.lean

```diff
@@ -0,0 +1,384 @@
+  e ← get_env,
+  l ← e.mfilter (λ d,
+    if e.in_current_file' d.to_name ∧ ¬ d.to_name.is_internal ∧ ¬ d.is_auto_generated e
+    then bnot <$> has_attribute' `nolint d.to_name else return ff),
+  lint_aux l (λ t, print_decls <$> fold_over_with_cond l t)
+    "in the current file" slow checks
+
+/-- Return the message printed by `#lint_mathlib`. -/
+meta def lint_mathlib (slow : bool := tt) (extra : list name := [])
+  (use_only : bool := ff) : tactic format := do
+  checks ← get_checks slow extra use_only,
+  e ← get_env,
+  ml ← get_mathlib_dir,
+  /- note: we don't separate out some of these tests in `lint_aux` because that causes a
+    performance hit. That is also the reason for the current formulation using if then else. -/
+  l ← e.mfilter (λ d,
+    if e.is_prefix_of_file ml d.to_name ∧ ¬ d.to_name.is_internal ∧ ¬ d.is_auto_generated e
+    then bnot <$> has_attribute' `nolint d.to_name else return ff),
+  let ml' := ml.length,
+  lint_aux l (λ t, print_decls_sorted_mathlib ml' <$> fold_over_with_cond_sorted l t)
+    "in mathlib (only in imported files)" slow checks
+
+/-- Return the message printed by `#lint_all`. -/
+meta def lint_all (slow : bool := tt) (extra : list name := [])
+  (use_only : bool := ff) : tactic format := do
+  checks ← get_checks slow extra use_only,
+  e ← get_env,
+  l ← e.mfilter (λ d, if ¬ d.to_name.is_internal ∧ ¬ d.is_auto_generated e
+    then bnot <$> has_attribute' `nolint d.to_name else return ff),
+  lint_aux l (λ t, print_decls_sorted <$> fold_over_with_cond_sorted l t)
+    "in all imported files (including this one)" slow checks
+
+/-- Parses an optional `only`, followed by a sequence of zero or more identifiers.
+Prepends `linter.` to each of these identifiers. -/
+private meta def parse_lint_additions : parser (bool × list name) :=
+prod.mk <$> only_flag <*> (list.map (name.append `linter) <$> ident_*)
+
+/-- The common denominator of `lint_cmd`, `lint_mathlib_cmd`, `lint_all_cmd` -/
+private meta def lint_cmd_aux (scope : bool → list name → bool → tactic format) : parser unit :=
+do b ← optional (tk "-"),
+   (use_only, extra) ← parse_lint_additions,
+   s ← scope b.is_none extra use_only,
+   trace s
+
+/-- The command `#lint` at the bottom of a file will warn you about some common mistakes
+in that file. Usage: `#lint`, `#lint linter_1 linter_2`, `#lint only linter_1 linter_2`.
+Use the command `#list_linters` to see all available linters. -/
+@[user_command] meta def lint_cmd (_ : parse $ tk "#lint") : parser unit :=
+lint_cmd_aux @lint
+
+/-- The command `#lint_mathlib` checks all of mathlib for certain mistakes.
+Usage: `#lint_mathlib`, `#lint_mathlib linter_1 linter_2`, `#lint_mathlib only linter_1 linter_2`.
+Use the command `#list_linters` to see all available linters. -/
+@[user_command] meta def lint_mathlib_cmd (_ : parse $ tk "#lint_mathlib") : parser unit :=
+lint_cmd_aux @lint_mathlib
+
+/-- The command `#lint_mathlib` checks all of mathlib for certain mistakes.
```

Comment:
```suggestion
/-- The command `#lint_all` checks all of mathlib for certain mistakes.
```

Reply:
@fpvandoorn the name `#lint_all` might suggest that `#lint` will run on the current project + all its dependencies. Is that intended? If not, I would prefer the more explicit `#lint_mathlib`.

Reply:
@jcommelin We have both `#lint_mathlib` and `#lint_all`. I copied the doc string from one to the other but only partially updated it.